[SYCL] Add ONEAPI_DEVICE_SELECTOR implementation (#6779)
intel/llvm: sycl/doc/EnvironmentVariables.md

| Environment variable | Values | Description |
| -------------------- | ------ | ----------- |

> | `ONEAPI_DEVICE_SELECTOR` | [See below.](#oneapi_device_selector)  | This device selection environment variable can be used to limit the choice of devices available when the SYCL-using application is run.  Useful for limiting devices to a certain type (like GPUs or accelerators) or backends (like LevelZero or OpenCL).  This device selection mechanism is replacing `SYCL_DEVICE_FILTER` .  The `ONEAPI_DEVICE_SELECTOR` syntax is shared with OpenMP and also allows sub-devices to be chosen. [See below.](#oneapi_device_selector) for a full description.   |

Reviewer: ```suggestion
| `ONEAPI_DEVICE_SELECTOR` | [See below.](#oneapi_device_selector)  | This device selection environment variable can be used to limit the choice of devices available when the SYCL-using application is run.  Useful for limiting devices to a certain type (like GPUs or accelerators) or backends (like Level Zero or OpenCL).  This device selection mechanism is replacing `SYCL_DEVICE_FILTER` .  The `ONEAPI_DEVICE_SELECTOR` syntax is shared with OpenMP and also allows sub-devices to be chosen. [See below.](#oneapi_device_selector) for a full description.   |
```

Search for "LevelZero" in this file (no space) and add a space.  There are just a few occurrences of that spelling.  In all other places we have the space.

Author: there were three of them. Got them.

A value for this environment value consists of a backend name, followed by a colon ( : ), followed by some sort of device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wilcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 

Possible values of `backend` are:
- `level_zero`
- `opencl`
- `cuda`
- `hip`
- `esimd_emulator`
- `*`

Possible named device types are:
- `cpu`
- `gpu`
- `acc`
- `*`

To see the exact numeric device id of each of the available devices, simply run `sycl-ls`.

Example: `ONEAPI_DEVICE_SELECTOR=opencl:2.0` This would select the first tile of some device with an id of 2 from OpenCL platform. 

Common Usage Examples:

| Example | Result |
-----------|---------
| `ONEAPI_DEVICE_SELECTOR=opencl:*` | All the devices on the OpenCL platform are available |
| `ONEAPI_DEVICE_SELECTOR=level_zero:gpu` | Only gpu devices on the LevelZero platform will be available.|

> | `ONEAPI_DEVICE_SELECTOR="opencl:gpu;level_zero:gpu"` | GPU devices from both Level Zero and OpenCL will be available. (Quotation marks will likely be needed when using semi-colon separated entries.) |

Reviewer 2: I think quotes are not needed in the parsed string itself -- at least not with the Ubuntu 20.04 WSL shell I tried -- although some form of shell escaping is needed.

```
holmesda@holmesda-mobl:~$ export ONEAPI_DEVICE_SELECTOR=level-zero:gpu;opencl:gpu
opencl:gpu: command not found
holmesda@holmesda-mobl:~$ export ONEAPI_DEVICE_SELECTOR='level-zero:gpu;opencl:gpu'
holmesda@holmesda-mobl:~$ echo $ONEAPI_DEVICE_SELECTOR
level-zero:gpu;opencl:gpu
holmesda@holmesda-mobl:~$ export ONEAPI_DEVICE_SELECTOR=level-zero:gpu\;opencl:gpu
holmesda@holmesda-mobl:~$ echo $ONEAPI_DEVICE_SELECTOR
level-zero:gpu;opencl:gpu
```

`(*) Note: Any means this environment variable is effective when set to any non-null value.`

### `ONEAPI_DEVICE_SELECTOR`

With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default choice will be one of them, usually preferring a LevelZero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit that choice of devices, and to expose sub-devices (aka "tiles") as individual devices.  

> A value for this environment value consists of a backend name, followed by a colon ( : ), followed by a device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wildcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 

Author: grammar has been added.

> A value for this environment value consists of a backend name, followed by a colon ( : ), followed by a device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wildcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 

Reviewer: > Sub-sub devices (ie two dots) are supported.

What do you mean by "supported"?  Did you verify that using two dots exposes each CCS as a root device in SYCL?  When I talked with @smaslov-intel a few days ago, he said that the PI layer didn't support partitioning a tile (sub-device) into CCS's (sub-sub-devices) yet.  Has that changed?

> A value for this environment value consists of a backend name, followed by a colon ( : ), followed by a device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wildcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 

Reviewer: > grammar has been added.

Thanks for adding the grammar.  I was thinking that we could use the grammar instead of the less precise wording, though.  I think our audience will understand BNF grammar.  Many Linux manpages use this grammar to describe the parameters that are accepted by commands.

Could we rewrite the description along these lines?

> With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default device selector will choose one of these devices, usually preferring a Level Zero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit the devices that are available to the application, and it can be used to expose GPU sub-devices (aka "tiles") as SYCL root devices.
>
> The syntax of this environment variable follows this BNF grammar:
>
> ```
> ONEAPI_DEVICE_SELECTOR = <selector-string>
> <selector-string> ::= <term>[;<term>...]
> <term> ::= <backend>:<devices>
> <backend> ::= { * | level_zero | opencl | cuda | hip | esimd_emulator }  // case insensitive
> <devices> ::= <device>[,<device>...]
> <device> ::= { * | cpu | gpu | fpga | <num> | <num>.<num> | <num>.* | *.* }  // case insensitive
> ```
>
> Each term in the grammar selects a collection of devices from a particular backend.  The device names `cpu`, `gpu`, and `fpga` select all devices from that backend with the corresponding type.  A backend's device can also be selected by its numeric index (zero-based) or by using `*` which selects all devices in the backend.
>
> The dot syntax (e.g. `<num>.<num>`) causes one or more GPU sub-devices (aka tiles) to be exposed to the application as SYCL root devices.  For example, `1.0` exposes the first tile of the second device as a SYCL root device.  The syntax `<num>.*` exposes all sub-devices of the give device as SYCL root devices.  The syntax `*.*` exposes all sub-devices of all GPU devices as SYCL root devices.
>
> In general, a term with one or more asterisks (`*`) matches all backends, devices, or sub-devices with the given pattern.  However, a warning is generated if the term does not match anything.  For example, `*:gpu` matches all GPU devices in all backends (ignoring backends with no GPU devices), but it generates a warning if there are no GPU devices in any backend.  Likewise, `level_zero:*.*` matches all sub-devices of partitionable GPUs in the Level Zero backend, but it generates a warning if there are no Level Zero GPU devices that are partitionable into sub-devices.
>
> The device indices are zero-based and are unique only within a backend.  Therefore, `level_zero:0` is a different device from `cuda:0`.  To see the indices of all available devices, run the `sycl-ls` tool.  Note that different backends sometimes expose the same hardware as different "devices".  For example, the `level_zero` and `opencl` backends both expose the Intel GPU devices.
>
> The following examples further illustrate the usage of this environment variable:
>
> | Example | Result |
> -----------|---------
> | `ONEAPI_DEVICE_SELECTOR=opencl:*` | Only the OpenCL devices are available |
> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu` | Only GPU devices on the Level Zero backend are available.|
> | `ONEAPI_DEVICE_SELECTOR="opencl:gpu;level_zero:gpu"` | GPU devices from both Level Zero and OpenCL are available. (Note that escaping (like quotation marks) will likely be needed when using semi-colon separated entries.) |
> | `ONEAPI_DEVICE_SELECTOR=opencl:gpu,cpu` | Only CPU and GPU devices on the OpenCL backend are available.|
> | `ONEAPI_DEVICE_SELECTOR=opencl:0` | Only the device with index 0 on the OpenCL backend is available. |
> | `ONEAPI_DEVICE_SELECTOR=hip:0,2` | Only devices with indices of 0 and 2 from the HIP backend are available. |
> | `ONEAPI_DEVICE_SELECTOR=opencl:0.*` | All the sub-devices (aka tiles) from the OpenCL device with index 0 are exposed as SYCL root devices.  No other devices are available. |
> | `ONEAPI_DEVICE_SELECTOR=opencl:0.2` | The third tile (2 in zero-based counting) of the OpenCL device with index 0 will be the sole device available.  |
> | `ONEAPI_DEVICE_SELECTOR=level_zero:*,*.*` | Exposes Level Zero devices to the application in two different ways.  Each device (aka "card") is exposed as a SYCL root device and each sub-device (aka "tile") is also exposed as a SYCL root device. |
>
> Notes:
> - The backend argument is always required. An error will be thrown if it is absent.
> - Additionally, the backend MUST be followed by colon ( : ) and at least one device specifier of some sort, else an error is thrown.
> - For sub-devices/tiles, the parent device must support partitioning by `info::partition_property::partition_by_affinity_domain` and `info::partition_affinity_domain::next_partitionable`. See the SYCL 2020 specification for a precise definition.)
> - The semi-colon character ( ; ) is treated specially by many shells, so you may need to enclose the string in quotes if the selection string contains this character. 

I omitted any description of the two-dot syntax here because I'm not sure if it's fully supported.  (See my other comment.)  We can add that back if we decide that it is.

> A value for this environment value consists of a backend name, followed by a colon ( : ), followed by a device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wildcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 

Author: @gmlueck  - yes, I tested it on some internal dev hardware and it worked. I was able to partition to sub-devices, and in-turn, partition those to sub-sub-devices, but no further.

Common Usage Examples:

| Example | Result |
-----------|---------
| `ONEAPI_DEVICE_SELECTOR=opencl:*` | All the devices on the OpenCL platform are available |
| `ONEAPI_DEVICE_SELECTOR=level_zero:gpu` | Only gpu devices on the LevelZero platform will be available.|
| `ONEAPI_DEVICE_SELECTOR="opencl:gpu;level_zero:gpu"` | GPU devices from both Level Zero and OpenCL will be available. (Note that escaping (like quotation marks) will likely be needed when using semi-colon separated entries.) |
| `ONEAPI_DEVICE_SELECTOR=opencl:gpu,cpu` | Only cpu and gpu devices on the OpenCL platform will be available.|
| `ONEAPI_DEVICE_SELECTOR=opencl:0` | Device with numeric identifier of 0 chosen from the OpenCL platform. |
| `ONEAPI_DEVICE_SELECTOR=hip:0,2` | Devices with numeric identifier of 0 and 2 are chosen from the HIP platform. |
| `ONEAPI_DEVICE_SELECTOR=opencl:0.*` | All the sub-devices (aka tiles) from the OpenCL device identified by 0 will be available. |
| `ONEAPI_DEVICE_SELECTOR=opencl:0.2` | The third tile (2 in 0 based counting) of the OpenCL device with numeric identifier 0 will be the sole device available.  |

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Author: >  All the devices selected by ONEAPI_DEVICE_SELECTOR will be SYCL root devices, so none of them will have a parent device.

That is definitely not what is happening. The `ONEAPI_DEVICE_SELECTOR`  limits the available choice of devices, and in the case of sub-devices, exposes partitions in the listing of devices gotten.  They are not modified in any way. So while a partition of some parent device may now be gotten via `device::get_devices()` or `platform::get_devices()`, which is new,  the get_info() applied against those devices would not be modified. If a device is a partition, I presume it'll have a parent, which the users code could get. 

It's cool and all that ONEAPI_DEVICE_SELECTOR can help free users from the burden of having to go and root around partitioning before they can even run a kernel, but pretending that a partition is not actually a partition seems ill-advised.

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Author: I made a gist that demos this (needs this PR for the ONEAPI_DEVICE_SELECTOR support, and a machine with a partitionable GPU): https://gist.github.com/cperkinsintel/a14cdf1666c474f428cbdbc5ab98272b

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Reviewer: >gpu.* selects not all the GPUs, but all partitions of any partitionable GPU.

This is also what `*.*` does, which is why we dropped `gpu.*`.

> That is definitely not what is happening. The ONEAPI_DEVICE_SELECTOR limits the available choice of devices, and in the case of sub-devices, exposes partitions in the listing of devices gotten. They are not modified in any way. So while a partition of some parent device may now be gotten via device::get_devices() or platform::get_devices(), which is new, the get_info() applied against those devices would not be modified. If a device is a partition, I presume it'll have a parent, which the users code could get.

This is contrary to the SYCL specification.  The specification of `device::get_devices()` is:

> Returns a `std::vector` containing all the *root devices* from all SYCL backends available in the system
which have the device type encapsulated by `deviceType`.

Note that it specifically says "root devices", which is defined in the glossary as:

> A device that is **not a sub-device**. The function `device::get_devices()` returns a vector of all the root devices.

And the spec says that devices that are not a sub-device do not  have any parent.  Here is the specification for `info::device::parent_device`:

> Returns the parent SYCL device to which this sub-device is a child if this is a sub-device.  **Must throw an exception with the `errc::invalid` error code if this SYCL device is not a sub device.**

The intention of the "dot" syntax is to allow programmers pretend that each tile is an independent root device.  Therefore, we should maintain this illusion and not expose any parent device.

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Author: @gmlueck  `*.*` is any partitionable device ( caveat affinity doman, next_partitionable) ,whereas `gpu.*` is any partitionable GPU (same caveats), correct? The fact may be that today those two sets are equal, but in the future, they might not be. For example, it is conceivable that a CPU device could be support such partitioning.

In any case, both are consistent with the BNF, no?

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Author: > The intention of the "dot" syntax is to allow programmers pretend that each tile is an independent root device. Therefore, we should maintain this illusion and not expose any parent device.

So you are saying that if the user specifies `ONEAPI_DEVICE_SELECTOR=level_zero:0,0.0` which would select two exact devices (the root device and one partition of it).   That we should NOT let the disambiguate those two devices from each other?  Something they would be able to do if they didn't use `ONEAPI_DEVICE_SELECTOR`  ?  To my mind, we are treading out of the "enable the programmer" zone by helping them into the "limit the programmer" by pretending we know better than them.  How on earth are we helping anyone by changing the value of  `get_info<info::device::parent>()` from it's real value to a pretend one?

> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu,gpu.*,gpu.*.*` | A wide net is cast. All level_zero gpus will be availabe in the device list, along with all their child partitions (if any), as well as the next level of partitions below them.  It would be up to the developer to figure out which device is which, if that is important. (This can be done with `info::device::parent_device`)|

Author: I see it as a tension between the spec and `ONEAPI_DEVICE_SELECTOR` which is, obviously, outside the specification.  In the strictest sense, what ONEAPI_DEVICE_SELECTOR proposes to do via sub-devices is a violation of the spec.  To my mind, that's is it's utility, it's why we are developing it. The boundaries the specification put up now make it difficult and tedious for users to work with partitions. ONEAPI_DEVICE_SELECTOR eases that, but its use exceeds the spec as it exists now, and that seems OK, to me at least.

Notes:
- The backend argument is always required. An error will be thrown if it is absent.
- Additionally, the backend MUST be followed by colon ( : ) and at least one device specifier of some sort, else an error is thrown.

> - For sub-devices/tiles, the parent device must support partitioning ( see`info::partition_by_domain` and `info::next_partitionable` in the SYCL 2020 Spec)

Reviewer: ```suggestion
- For sub-devices/tiles, the parent device must support partitioning into sub-devices by `info::info::partition_property::partition_by_affinity_domain` and `info::info::partition_affinity_domain::next_partitionable`.  See the SYCL 2020 specification for a precise definition.
```

> - if using semi-colons ( ; ) to separate entries, quotation marks will likely be needed around the entire environment variable. 

Reviewer: I think it would be better to say something like:

> The semi-colon character (`;`) is treated specially by many shells, so you may need to enclose the string in quotes if the selection string contains this character.

`(*) Note: Any means this environment variable is effective when set to any non-null value.`

### `ONEAPI_DEVICE_SELECTOR`

> With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default choice will be one of these devices, usually preferring a LevelZero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit that choice of devices, and to expose GPU sub-devices (aka "tiles") as individual devices.  

Reviewer: ```suggestion
With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default choice will be one of these devices, usually preferring a LevelZero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit that choice of devices, and to expose GPU sub-devices or sub-sub-devices as individual devices.  
```

I recently learned that the sub-devices are not necessarily tiles, so we should remove all the wording which equates "sub-device"  with "tile".  We should just call them sub-devices.

Search for "tile" in the wording below and make similar changes.

The syntax of this environment variable follows this BNF grammar:
```
ONEAPI_DEVICE_SELECTOR = <selector-string>
<selector-string> ::= <term>[;<term>...]
<term> ::= <backend>:<devices>
<backend> ::= { * | level_zero | opencl | cuda | hip | esimd_emulator }  // case insensitive
<devices> ::= <device>[,<device>...]

> <device> ::= { * | cpu | gpu | fpga | <num> | <num>.<num> | <num>.* | *.* }  // case insensitive

Reviewer: Please add the sub-sub-device grammar here.  I assume it is like this?

```suggestion
<device> ::= { * | cpu | gpu | fpga | <num> | <num>.<num> | <num>.* | *.* | <num>.<num>.<num> | <num>.<num>.* | <num>.*.* | *.*.* }  // case insensitive
```

```

> Each term in the grammar selects a collection of devices from a particular backend. The device names cpu, gpu, and fpga select all devices from that backend with the corresponding type. A backend's device can also be selected by its numeric index (zero-based) or by using * which selects all devices in the backend.

Reviewer: ```suggestion
Each term in the grammar selects a collection of devices from a particular backend. The device names cpu, gpu, and fpga select all devices from that backend with the corresponding type. A backend's device can also be selected by its numeric index (zero-based) or by using `*` which selects all devices in the backend.
```

In general, a term with one or more asterisks ( `*` ) matches all backends, devices, or sub-devices with the given pattern. However, a warning is generated if the term does not match anything. For example, `*:gpu` matches all GPU devices in all backends (ignoring backends with no GPU devices), but it generates a warning if there are no GPU devices in any backend. Likewise, `level_zero:*.*` matches all sub-devices of partitionable GPUs in the Level Zero backend, but it generates a warning if there are no Level Zero GPU devices that are partitionable into sub-devices.

The device indices are zero-based and are unique only within a backend. Therefore, `level_zero:0` is a different device from `cuda:0`. To see the indices of all available devices, run the `sycl-ls` tool. Note that different backends sometimes expose the same hardware as different "devices". For example, the level_zero and opencl backends both expose the Intel GPU devices.

> Additionally, if a sub-device is chosen (via numeric index or wildcard), then an additional layer of partitioning can be specified. In other words, a sub-sub-device can be selected. Like sub-devices, this is done with a period ( `.` ) and a sub-sub-device specifier which is a wildcard symbol ( `*` ) or a numeric index.  Example `ONEAPI_DEVICE_SELECTOR=level_zero.0.*.*` would partition device 0 into sub-devices and then partition each of those into sub-sub-devices. The range of grandchild sub-sub-devices would be the final devices available to the app, neither device 0, nor its child partitions would be in that list. 

Reviewer: ```suggestion
Additionally, if a sub-device is chosen (via numeric index or wildcard), then an additional layer of partitioning can be specified. In other words, a sub-sub-device can be selected. Like sub-devices, this is done with a period ( `.` ) and a sub-sub-device specifier which is a wildcard symbol ( `*` ) or a numeric index.  Example `ONEAPI_DEVICE_SELECTOR=level_zero:0.*.*` would partition device 0 into sub-devices and then partition each of those into sub-sub-devices. The range of grandchild sub-sub-devices would be the final devices available to the app, neither device 0, nor its child partitions would be in that list. 
```

Possible values of `backend` are:
- `level_zero`
- `opencl`
- `cuda`
- `hip`
- `esimd_emulator`
- `*`

Possible named device types are:
- `cpu`
- `gpu`
- `fpga`
- `*`

To see the exact numeric device id of each of the available devices, simply run `sycl-ls`.

> 

Reviewer: I would delete all of this because it is already described above.

| `ONEAPI_DEVICE_SELECTOR=opencl:gpu,cpu` | Only CPU and GPU devices on the OpenCL platform are available.|
| `ONEAPI_DEVICE_SELECTOR=opencl:0` | Only the device with index 0 on the OpenCL backend is available. |
| `ONEAPI_DEVICE_SELECTOR=hip:0,2` | Only devices with indices of 0 and 2 from the HIP backend are available. |
| `ONEAPI_DEVICE_SELECTOR=opencl:0.*` | All the sub-devices (aka tiles) from the OpenCL device with index 0 are exposed as SYCL root devices. No other devices are available. |
| `ONEAPI_DEVICE_SELECTOR=opencl:0.2` | The third tile (2 in zero-based counting) of the OpenCL device with index 0 will be the sole device available.  |
| `ONEAPI_DEVICE_SELECTOR=level_zero:*,*.*` | Exposes Level Zero devices to the application in two different ways. Each device (aka "card") is exposed as a SYCL root device and each sub-device (aka "tile") is also exposed as a SYCL root device.|

Notes:
- The backend argument is always required. An error will be thrown if it is absent.
- Additionally, the backend MUST be followed by colon ( : ) and at least one device specifier of some sort, else an error is thrown.

> - For sub-devices/tiles, the parent device must support partitioning ( see`info::partition_property::partition_by_affinity_domain` and `info::partition_affinity_domain::next_partitionable`. See the SYCL 2020 specification for a precise definition.)

Reviewer: ```suggestion
- For sub-devices and sub-sub-devices, the parent device must support partitioning ( see`info::partition_property::partition_by_affinity_domain` and `info::partition_affinity_domain::next_partitionable`. See the SYCL 2020 specification for a precise definition.)  For Intel GPUs, the sub-device and sub-sub-device syntax can be used to expose tiles or CCSs to the SYCL application as root devices.  The exact mapping between sub-device, sub-sub-device, tiles, and CCSs is specific to the hardware.
```

We can use this note to explain that sub-devices and sub-sub-devices map to tiles and CCSs.
